src/arch/x86/64/c_traps.c

```diff
@@ -39,21 +39,73 @@ static void NORETURN restore_vmx(void)
     /* Do not support breakpoints in VMs, so just disable all breakpoints */
     loadAllDisabledBreakpointState(cur_thread);
 #endif
+    vcpu_restore_guest_msrs(cur_thread->tcbArch.tcbVCPU);
     if (cur_thread->tcbArch.tcbVCPU->launched) {
         /* attempt to do a vmresume */
         asm volatile(
-            // Set our stack pointer to the top of the tcb so we can efficiently pop
-            "movq %[reg], %%rsp\n"
+            // Arguments are getting stored in general purpose registers that need to be used.
+            // Copy them to unused general purpose registers
+            "movq %[host_msr], %%r8\n"
+            "movq %[guest_msr], %%r9\n"
+            "movq %[reg], %%r10\n"
+
+#ifdef ENABLE_SMP_SUPPORT
+            // Save host's GS, Shadow GS, and FS
+            "mov $0xC0000101, %%ecx\n"
+            "rdmsr\n"
+            "shl $0x20,%%rdx\n"
+            "or %%rdx, %%rax\n"
+            "mov %%rax, %%r11\n" // R11 has GS
+            "swapgs\n"
```

Comment:
maybe add a comment to indicate that swapgs swaps two specific MSRs (IA32_GS_BASE and IA32_KERNE_GS_BASE).